### PR TITLE
Feature/check_dataset_comp

### PR DESCRIPTION
**Pull request by KaiWaldrant**
Add a component to check the dataset schema defined in the api anndata_*.yaml files.

### CODE REVIEW

File: src/common/check_dataset_schema/script.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    out['error'][slot] = check
+
+if par['stop_on_error'] == 'true':
+  write_json(out)
+  exit(out['exit_code'])
+
+if par['copy_output'] == 'true':
+  assert par['output'] is not None, 'No output defined'
+  write_json(out)
+  shutil.copyfile(par["input"], par["output"])
```
Comment (KaiWaldrant) [inline src/common/check_dataset_schema/script.py:69]: thinking about using stop_on_error as argument to check if output needs to be created. and remove the copy_output arg. 

If you are not exiting on error then for instance nextflow will not have a exit code 1 if it failed or an output if everything was ok. if `stop_on_error` is false then the exit code will be `0` regardless if the dataset was correct or not.

File: src/common/check_dataset_schema/script.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+}
+
+for slot in def_slots:
+  check = check_structure(def_slots[slot], getattr(adata, slot))
+  if bool(check):
+    out['exit_code'] = 1
+    out['data_schema'] = 'not ok'
+    out['error'][slot] = check
+
+if par['stop_on_error'] == 'true':
```
Comment (rcannood) [inline src/common/check_dataset_schema/script.py:62]:
```suggestion
if par['stop_on_error']:
```

stop_on_error is not a string.

File: src/common/check_dataset_schema/script.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
+import anndata as ad
+import yaml
+import shutil
+import json
+
+
+## VIASH START
+# The following code has been auto-generated by Viash.
```
Comment (rcannood) [inline src/common/check_dataset_schema/script.py:8]:
```suggestion
```

File: src/common/check_dataset_schema/script.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+  'stop_on_error': 'false',
+  'copy_output': 'false',
+  'json': 'output/error.json',
+  'output': 'output/output.h5ad'
+}
+meta = {
+  'functionality_name': 'foo',
+
+}
+
```
Comment (rcannood) [inline src/common/check_dataset_schema/script.py:21]:
```suggestion
```

File: src/common/check_dataset_schema/script.py (hunk 1 above)
Comment (rcannood) [inline src/common/check_dataset_schema/script.py:69]: Simplify code according to suggestion above

```suggestion
if par['json'] is not None:
  with open(par["checks"], "w") as f:
    json.dump(out, f, indent=2)

if par['output'] is not None:
  shutil.copyfile(par["input"], par["output"])
    
if par['stop_on_error']:
  exit(out['exit_code'])
```

File: src/common/check_dataset_schema/config.vsh.yaml [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+          description: A schema file for the h5ad object.
+    - name: Arguments
+      arguments:
+        - name: --stop_on_error
+          type: boolean
+          default: false
+          description: Whether or not to stop with exit code 1 if the input file does not adhere to the schema.
+        - name: --copy_output
+          type: boolean_true
+          description: Wether a output needs to be created for further analysis.
```
Comment (rcannood) [inline src/common/check_dataset_schema/config.vsh.yaml:24]:
```suggestion
```

File: src/common/check_dataset_schema/config.vsh.yaml [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    - name: Output
+      arguments:
+        - name: --json
+          type: file
+          required: false
+          description: If specified, this file will contain a structured log of which checks succeeded (or not).
+        - name: --output
+          type: file
+          required: false
+          description: If specified, the output file will be a copy of the input file.
```
Comment (rcannood) [inline src/common/check_dataset_schema/config.vsh.yaml:34]:
```suggestion
        - name: --checks
          type: file
          required: false
          description: If specified, this file will contain a structured log of which checks succeeded (or not).
          example: checks.json
        - name: --output
          type: file
          required: false
          description: If specified, the output file will be a copy of the input file.
          example: output.h5ad
```

**Review by rcannood — CHANGES_REQUESTED** [review]

File: src/common/check_dataset_schema/script.py (hunk 2 above)
Comment (KaiWaldrant) [reply src/common/check_dataset_schema/script.py:62]: I know but for some reason this was not working when using `viash run` but seems to work now. probaly did something wrong back then

**Comment by rcannood** [comment]
Hmm.. How about we put the code that creates the test resource files inside the script? That way we don't need to sync test resources to get the test to work. Wdyt?

**Comment by KaiWaldrant** [comment]
fine with me, I'll change it.

File: src/common/check_dataset_schema/test.py [hunk 7]
```diff
@@ -3,12 +3,54 @@
 import json
 
 input_path = meta["resources_dir"] + "resources_test/common/pancreas/dataset.h5ad"
-input_correct_schema = meta["resources_dir"] +  "resources_test/common/check_schema/anndata_correct.yaml"
-input_error_schema = meta["resources_dir"] + "resources_test/common/check_schema/anndata_error.yaml"
+input_correct_schema = meta["resources_dir"] +  "resources_test/common/anndata_correct.yaml"
+input_error_schema = meta["resources_dir"] + "resources_test/common/anndata_error.yaml"
```
Comment (rcannood) [inline src/common/check_dataset_schema/test.py:7]: No need to use the `meta["resources_dir"]`

```suggestion
input_correct_schema = "anndata_correct.yaml"
input_error_schema = "anndata_error.yaml"
```

**Review by rcannood — CHANGES_REQUESTED** [review]

File: src/common/check_dataset_schema/config.vsh.yaml [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+          example: output.h5ad
+          direction: output
+  resources:
+    - type: python_script
+      path: script.py
+  test_resources:
+    - path: ../../../resources_test
+    - path: ../../../src
+      dest: src
```
Comment (rcannood) [inline src/common/check_dataset_schema/config.vsh.yaml:42]: Is `src/` still needed? We can be a little more specific on the test resources, too:

```suggestion
    - path: ../../../resources_test/common/pancreas
```

File: src/common/check_dataset_schema/test.py [hunk 9]
```diff
@@ -0,0 +1,100 @@
+import subprocess
+from os import path
+import json
+
+input_path = meta["resources_dir"] + "resources_test/common/pancreas/dataset.h5ad"
```
Comment (rcannood) [inline src/common/check_dataset_schema/test.py:5]: If the other suggestion gets accepted.

```suggestion
input_path = meta["resources_dir"] + "/pancreas/dataset.h5ad"
```

**Review by rcannood — CHANGES_REQUESTED** [review]

**Review by rcannood — APPROVED** [review]

**Pull request merged**